internal/pkg/build/oci/oci.go

```diff
@@ -88,38 +98,51 @@ func (t *ImageReference) newImageSource(ctx context.Context, sys *types.SystemCo
 // ParseImageName parses a uri (e.g. docker://ubuntu) into it's transport:reference
 // combination and then returns the proper reference
 func ParseImageName(ctx context.Context, imgCache *cache.Handle, uri string, sys *types.SystemContext) (types.ImageReference, error) {
-	ref, err := parseURI(uri)
+	ref, _, err := parseURI(uri)
 	if err != nil {
 		return nil, fmt.Errorf("unable to parse image name %v: %v", uri, err)
 	}
 
 	return ConvertReference(ctx, imgCache, ref, sys)
 }
 
-func parseURI(uri string) (types.ImageReference, error) {
+func parseURI(uri string) (types.ImageReference, string, error) {
 	sylog.Debugf("Parsing %s into reference", uri)
 
+	arch := getArchFromURI(uri)
+
 	split := strings.SplitN(uri, ":", 2)
 	if len(split) != 2 {
-		return nil, fmt.Errorf("%s not in transport:reference pair", uri)
+		return nil, arch, fmt.Errorf("%s not in transport:reference pair", uri)
 	}
 
 	transport := transports.Get(split[0])
 	if transport == nil {
-		return nil, fmt.Errorf("%s not a registered transport", split[0])
+		return nil, arch, fmt.Errorf("%s not a registered transport", split[0])
 	}
 
-	return transport.ParseReference(split[1])
+	imgRef, err := transport.ParseReference(split[1])
+	return imgRef, arch, err
 }
 
 // ImageDigest obtains the digest of a uri's manifest
-func ImageDigest(ctx context.Context, uri string, sys *types.SystemContext) (string, error) {
-	ref, err := parseURI(uri)
+func ImageDigest(ctx context.Context, uri string, sys *types.SystemContext, pullArch string) (digest string, err error) {
+	ref, arch, err := parseURI(uri)
 	if err != nil {
 		return "", fmt.Errorf("unable to parse image name %v: %v", uri, err)
 	}
 
-	return getRefDigest(ctx, ref, sys)
+	digest, err = getRefDigest(ctx, ref, sys)
```

Comment:
`getRefDigest` should follow the new hash computation logic (digest+arch), currently it looks like only pull will use this new digest, build will keep using old digest leading to cache miss

Reply:
thanks! I made changes in the latest commit
https://github.com/apptainer/apptainer/pull/1295/commits/470ee6eaea3daaebd8135d5f3044926d0c5ebece